internal/ethapi/api_test.go

```diff
@@ -2202,7 +2202,14 @@ func TestRPCGetBlockReceipts(t *testing.T) {
 			t.Errorf("test %d: want no error, have %v", i, err)
 			continue
 		}
+		// confirm the result matches the expected output
 		testRPCResponseWithFile(t, i, result, "eth_getBlockReceipts", tt.file)
```

Comment:
I think it would be nicer to get access to the golden file content directly, and unmarshal the file content into the receipt struct. You can then marshal the receipt struct again into an indented JSON string, and assert that the marshaled receipt matches the golden file content.

For that, you could add an additional `any` parameter to `testRPCResponseWithFile` into which you unmarshal and at the calling site you instantiate it, like
```suggestion
		rec := new(types.Receipt)
		testRPCResponseWithFile(t, i, result, "eth_getBlockReceipts", tt.file, rec)
```
this would also work for the other callers of `testRPCResponseWithFile`.

Reply:
I've taken your advice and made the expected type part of the `testRPCResponseWithFile` parameters, so it can confirm unmarshalling.

Interestingly, however, when I added the check you mention here, that the `type.Whatever` marshals *back* to the same as the "golden file content", The `TestRPCGetBlockOrHeader` tests fail!

```
        	Messages:   	test 0: json does not match type struct, want: {
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,4 +1,6 @@
        	            	-(map[string]interface {}) (len=18) {
        	            	+(map[string]interface {}) (len=21) {
        	            	  (string) (len=13) "baseFeePerGas": (string) (len=9) "0xfdc7303",
        	            	+ (string) (len=11) "blobGasUsed": (interface {}) <nil>,
        	            	  (string) (len=10) "difficulty": (string) (len=7) "0x20000",
        	            	+ (string) (len=13) "excessBlobGas": (interface {}) <nil>,
        	            	  (string) (len=9) "extraData": (string) (len=2) "0x",
        	            	@@ -12,2 +14,3 @@
        	            	  (string) (len=6) "number": (string) (len=3) "0xa",
        	            	+ (string) (len=21) "parentBeaconBlockRoot": (interface {}) <nil>,
        	            	  (string) (len=10) "parentHash": (string) (len=66) "0xedb9ccf3a85f67c095ad48abfb0fa09d47179bb0f902078d289042d12428aca5",
        	            	@@ -17,4 +20,4 @@
        	            	  (string) (len=9) "timestamp": (string) (len=4) "0x64",
        	            	- (string) (len=15) "totalDifficulty": (string) (len=3) "0x1",
        	            	- (string) (len=16) "transactionsRoot": (string) (len=66) "0xb0893d21a4a44dc26a962a6e91abae66df87fb61ac9c60e936aee89c76331445"
        	            	+ (string) (len=16) "transactionsRoot": (string) (len=66) "0xb0893d21a4a44dc26a962a6e91abae66df87fb61ac9c60e936aee89c76331445",
        	            	+ (string) (len=15) "withdrawalsRoot": (interface {}) <nil>
        	            	 }
```

I've updated this PR to this version which has this issue. I'll have to look at it in the morning, but to me it indicates that not all `type` objects will look the same as their original API counterparts once unmarshalled+marshalled.

Reply:
I see. Looking at the definition of `types.Header`, no struct field has `omitempty` set in their `json` tag, so missing fields are marshaled as `null`s instead of omitted. It's probably an inconsistency that has just never been detected (or just ignored) because such a test didn't exist before.

`totalDifficulty` is a special case that's added by the API server that's not part of the `Header` struct, see `BlockChainAPI.rpcMarshalHeader` and `rpcMarshalBlock`.

I guess we could just skip headers and blocks for now by passing `nil` as `unmarshalTo` and track solving this, possibly by raising an PR upstream to add `omitempty` to those fields (and add special treatment for field `totalDifficulty`). We could actually also ask upstream why there even are these manual `rpcMarshal...` functions, if the same can be achieved with properly configured `json` Go field tags.

Reply:
Agreed, I've added `nil` to the Header/Block test